Readme
WCSTools 3.9.4

WCSTools is a set of software utilities, written in C, which create,
display and manipulate the world coordinate system of a FITS or IRAF
image, using specific keywords in the image header which relate pixel
position within the image to position on the sky.  Auxillary programs
search star catalogs and manipulate images.

To make the entire package, type

make all

in the main directory or, on Solaris machines

make -f Makefile.solaris

This is version 3.9.4 of the WCSTools package developed by Jessica Mink
(jmink@cfa.harvard.edu) at the Smithsonian Astrophysical Observatory,
with significant code or algorithms from

Mark Calabretta of CSIRO (projection),
Bill Cotton of NRAO (projection),
Jesse Doggett and Rick White of StScI (DSS projection)
Elwood Downey of the University of Iowa (WCS fitting procedure),
John Roll of SAO (network access),
Jing Li of IPAC (SIP distortion),
Pat Wallace of Rutherford Appleton Laboratory, UK (coordinate conversion),
Frank Valdes of NOAO (support for NOAO's ZPX projection), and
Ed Los of HCO (support for Emmanuel Bertin's SCAMP distortion correction).

Bill Joye and Eric Mandel of SAO, Allan Brighton of ESO, and Jean-Baptiste
Marquette of the Institut d'Astrophysique de Paris, and David Shupe of IPAC
helped debug the subroutine library.
Norman Grogin, Scott Kenyon, Susan Tokarz, Dan Koranyi, Guillermo Torres,
Steve Willner, Gus Muensch, Ed Los and others of the CfA and Andreas Jaunsen
helped debug the astrometry programs.
Sergio Pascual of the Universidad Complutense de Madrid helped me get
all of the variables initialized.

The update history is in the NEWS file, but here is what is new:

Version 3.9.4 (August 2, 2016) (after Ole Streicher)
getdate: Fix input list so it works correctly (2016-08-02 for Armin Rest)
imstar: Append to output line rather than overwriting it
bincatc: Fix mis-reading of coordinates from command line
cphead: Increase number of keywords to handle distortion polynomials
filext: If no file extension, print null string, avoiding seg fault
getfits: Decrement argument counter after reading center RA
imcat: Fix typo in UCAC output format
imstar: Fix sprintf of headline
findstar.c: Fix but in trim section parsing
fitsfile.c: Add one to header string length for trailing NULL
wcs.c: Clean up use of ptype so it is always 3 characters long
sortstar.c: Initialize haspm flag if zero
Update all man pages

Version 3.9.3 (June 23, 2016)
wcs.h, wcs.c: Increase ctype string length to 15 (+null=16) for distortion
wcslib.h, wcslib.c: Increase ctype string length to 15 (+null=16) for distortion
scat.c, ucacread.c: Typos in code fixed
matchstar.c, hget.c, fitsfile.c: Fix isnum() tests for added coloned times and dashed date

Version 3.9.2 (December 10, 2015)
isnum: Add return of 4 for yyyy-mm-dd dates (2015-11-06)
isnum: Add -n option to output number without linefeed
imcat, imstar: Print two decimal places on PROS region file coordinates
remap: Declare GetFITSWCS()
simpos: Switch to CfA Vizier server as default (2015-05-05)
simpos: Use NED name resolver if called as nedpos or using -n argument
simpos: Look up in VIZIER catalogs if called as vizpos or using -z argument
simpos: Use CDS server if called as cdspos or using -c argument
sky2xy: Add xy option to -o to print only computed x and y coordinates
nedpos,nedname: Drop tasks because IPAC has dropped interface; use simpos -n
getdate: Fix bug with FITS date conversion (2015-08-24)
filedir: Add option to read file pathnames from stdin
catutil.c: Add tab as an assignment character in agets()
hget.c: Add return of 4 for yyyy-mm-dd dates to isnum() (2015-11-06)

Documentation, with examples and installation directions, is on the
World Wide Web at

        http://tdc-www.harvard.edu/software/wcstools/

References

If the WCSTools package has proven useful in your work, please reference
at least one of the following papers. If you want to find out which is
most applicable, the full texts are available on the WCSTools web site at

        http://tdc-www.harvard.edu/software/wcstools/publications/

The paper presented at ADASS in 1996 is the best published description
of the IMWCS world coordinate system fitting program: 
     "WCSTools: Putting Image World Coordinate Systems to Use",
     Douglas J. Mink, 1997, in Astronomical Data Analysis Software and
     Systems VI, A.S.P. Conference Series, Vol. 125. Edited by Gareth Hunt
     and H. E. Payne.
     San Francisco: Astronomical Society of the Pacific, pp. 249-252. 

The 1998 ADASS paper is the best published description of the other tools
in the WCSTools package, including SAOimage when used for WCS work: 
     "WCSTools: An Image Astrometry Toolkit",
     Douglas J. Mink, 1999, in Astronomical Data Analysis Software and
     Systems VIII A.S.P. Conference Series, Vol. 172.  Edited by Dave
     Mehringer, Ray Plante, and Doug Roberts.
     San Francisco: Astronomical Society of the Pacific, pp.  498-501. 

A paper presented at the 2001 ADASS conference describes how the
WCSTools package uses the web, among other new features:
     "WCSTools 3.0: More Tools for Image Astrometry and Catalog Searching",
     Douglas J. Mink, in Astronomical Data Analysis Software and Systems XI,
     A.S.P. Conference Series, Vol. 281, 2002. Edited by David
     Bohlender, Daniel Durand, and T.H. Handley.
     San Francisco: Astronomical Society of the Pacific, pp. 169-172.

This paper presented at the 2002 ADASS conference describes interesting
ways to use the SCAT program:
     "Federating Catalogs and Interfacing Them with Archives: A VO Prototype",
     Douglas J. Mink and Michael J. Kurtz, 2003, in Astronomical Data Analysis
     Software and Systems XII, A.S.P. Conference Proceedings, Vol. 295.
     Edited by Harry Payne, Robert Jedrzejewski, and Richard Hook.
     San Francisco: Astronomical Society of the Pacific, pp. 51-54.

To assist users in selecting the best catalog to use, we tested the accuracy of
the various catalogs one is likely to use with imwcs, showing how well it works
in a pipeline mode on 1728 images. This ADASS paper is an improved version of
our AAS/DDA paper, adding the UCAC2 catalog and emphasizing usefulness more than
astrometry:
     "A Comparison of Large All-Sky Catalogs",
     Douglas J. Mink, Warren R. Brown, and Michael J. Kurtz, 2004, in
     Astronomical Data Analysis Software and Systems XIII, A.S.P. Conference
     Proceedings, Vol. 314. Edited by Francois Oschenbein, Mark Allen,
     and Daniel Egret.
     San Francisco: Astronomical Society of the Pacific, pp. 141-144. 

Program Descriptions

addpix	Add a constant value(s) to specified pixel(s)
bincat	Bin a catalog into a FITS image in flux or number
conpix	Operate on all of the pixels of an image
cphead	Copy keyword values between FITS or IRAF images
delhead	Delete specified keywords from FITS or IRAF image file headers
delwcs	Delete the WCS keywords from an image. If both EPOCH and EQUINOX
	are present, EQUINOX is also deleted. 
edhead	Edit the header of a FITS file or the user parameters of an IRAF
	image file using a the text editor specified by the EDITOR environment
	variable.
getcol	Extract specified fields from an space-separated ASCII table file
getdate Convert dates and times between various formats
getfits	Extract portion of a FITS file into a new FITS file, preserving WCS
gethead	Return values for keyword(s) specified after filename.
getpix	Return value(s) of specified pixel(s)
gettab	Extract values from tab table data base files
i2f	Read two-dimensional IRAF image file and write FITS image file
imcat	List catalog sources which should be found in the area of the sky
	covered by a specific image.
	Specific catalogs supported include the GSC II, the 2MASS Point
	Source Catalog, the HST Guide Star Catalog, the USNO-A2.0 Catalog,
	the Tycho-2 Catalog, the PPM Catalog, the IRAS Point Source Catalog,
	and the SAO Catalog.
imextract Extract 1- or 2-dimensional images from 2- or 3-dimensional images
imfill	 Replace bad pixels using a box median, mean, or Gaussian filter
imhead	Print FITS or IRAF header
immatch	Match catalog and image stars using the WCS in the image file.
imresize Shrink an image by an integral factor keeping WCS intact
imrot	Rotate and/or reflect FITS or IRAF image files
imsize	Print center and size of image using WCS keywords in header
imsmooth Run a box median, mean, or Gaussian filter over an image
imstack	Stack 1-dimensional images into a 2-dimensional image
imstar	Find and list the n brightest stars in an IRAF or FITS image, with
	their sky coordinates if there is WCS information in the image header. 
imwcs	Automaticaly find stars in a FITS or IRAF image, match them to any
	of the supported catalogs, compute the relation between sky coordinates
	and image coordinates, and write it into the image header.
keyhead	Change keyword names in headers of FITS or IRAF images.
newfits	Create blank FITS image files (dataless by default, with BITPIX=0)
remap	Remap an image from one WCS into another, rebinning as necessary
scat	Search a source catalog given a region on the sky.
	Specific catalogs supported include the GSC II, the USNO-B1.0 Catalog,
	the 2MASS Point Source Catalog, the HST Guide Star Catalog, the
	USNO-A2.0 Catalog, the USNO UCAC2 Catalog, the Tycho-2 Catalog,
	the PPM Catalog, the IRAS Point Source Catalog, and the SAO Catalog.
sethead	Set header keyword values in FITS or IRAF images.
setpix	Set specified pixel(s) to specified value(s)
simpos  Return RA and Dec for object name (or list of them) using SIMBAD
sky2xy	Print image pixel coordinates for given sky coordinates on the
	command line or in a list file. 
skycoor	Convert between J2000, B1950, and Galactic coordinates from the
	command line or a file.
sumpix	Total pixel values over an image row or column or a specified area
wcshead	Print basic world coordinate system information on one line per image
xy2sky	Print sky coordinates for given image pixel coordinates on the command
	line or in a list file.

-Jessica Mink, SAO, 2016-08-02
 http://tdc-www.harvard.edu/mink/


NEWS
WCSTools Package updates

Version 3.9.4 (August 2, 2016) (after Ole Streicher)
getdate: Fix input list so it works correctly (2016-08-02 for Armin Rest)
imstar: Append to output line rather than overwriting it
bincatc: Fix mis-reading of coordinates from command line
cphead: Increase number of keywords to handle distortion polynomials
filext: If no file extension, print null string, avoiding seg fault
getfits: Decrement argument counter after reading center RA
imcat: Fix typo in UCAC output format
imstar: Fix sprintf of headline
findstar.c: Fix but in trim section parsing
fitsfile.c: Add one to header string length for trailing NULL
wcs.c: Clean up use of ptype so it is always 3 characters long
sortstar.c: Initialize haspm flag if zero
Update all man pages

Version 3.9.3 (June 23, 2016)
wcs.h, wcs.c: Increase ctype string length to 15 (+null=16) for distortion
wcslib.h, wcslib.c: Increase ctype string length to 15 (+null=16) for distortion
scat.c, ucacread.c: Typos in code fixed
matchstar.c, hget.c, fitsfile.c: Fix isnum() tests for added coloned times and dashed date

Version 3.9.2 (December 10, 2015)
isnum: Add return of 4 for yyyy-mm-dd dates (2015-11-06)
isnum: Add -n option to output number without linefeed
imcat, imstar: Print two decimal places on PROS region file coordinates
remap: Declare GetFITSWCS()
simpos: Switch to CfA Vizier server as default (2015-05-05)
simpos: Use NED name resolver if called as nedpos or using -n argument
simpos: Look up in VIZIER catalogs if called as vizpos or using -z argument
simpos: Use CDS server if called as cdspos or using -c argument
sky2xy: Add xy option to -o to print only computed x and y coordinates
nedpos,nedname: Drop tasks because IPAC has dropped interface; use simpos -n
getdate: Fix bug with FITS date conversion (2015-08-24)
filedir: Add option to read file pathnames from stdin
catutil.c: Add tab as an assignment character in agets()
hget.c: Add return of 4 for yyyy-mm-dd dates to isnum() (2015-11-06)

Version 3.9.1 (March 24, 2015)
remap: Declare GetFITSWCS() as well as GetWCSFITS()
cphead: Fix bug freeing header string more than once
imcat,scat,immatch,imwcs: Fix bugs with web access to GSC2 and SDSS catalogs
gsc2read.c: Drop concatenation of "blank" string in middle of search URL
webread.c: Fix bug in URL filtering

Version 3.9.0 (December 19, 2014)
gethead: Add -r option to return multiple values as a single word separated
         by the first character of the keyword the value of which follows
         (2014-12-19)
remap: Update usage to -f WCSfile suggested by Steve Willner (2014-02-19)
sethead: Fix bug dealing with setting the same keyword in multiple file extensions
imcat, scat, immatch, imwcs: Implement UCAC4 catalog
fileutil.c: Add next_line() to return one line of file
fitfile.c: fix buffer reallocation bug in fitsrhead()
tabread.c: fix file tests
gsc2read.c and tabread.c: adjust to GSC2.3.3 format
webbuff.c: Add filter to convert characters in web URLs

Version 3.8.9 (September 16, 2013)
imcat, imwcs, immatch, scat: Work around SDSS server failure
sdssread.c: Add alternate UK and US catalog servers

Version 3.8.8 (April 30, 2013)
getdate: Add -s option to subtract (or add) days
getdate: Add self-conversion for *JD's
getfits: Print error, not blank file, if requested region is off image
simpos: Update URL for SIMBAD query

Version 3.8.7 (October 31, 2012)
Rename Man to man to avoid problems with Mac OSX file name convention
filedir: If pathname ends in "/", drop last directory
catutil.c:  Skip trailing right bracket in aget*()
dateutil.c:  Unused l0 dropped from jd2lst(); ts2ss from jd2mst()
imio.c:  Fix errors with short and character images in minvec(), maxvec()
wcs.c:  Drop d1 and d2 from wcsdist(); diffi from wcsdist1()
wcs.c:  Drop depwcs; it's in main wcs structure
wcsinit.c: Drop unused variable iszpx; fix bug in latmin assignment
zpxpos.c: Fix code for quadratic near pole

Version 3.8.6 (August 10, 2012)
sumpix: Add options i to print index and e to compute medians
gethead: Call fk425e() instead of fk425()
imcat: Drop declaration of unused fk425e()
imutil.c: Add medpix() and meanpix() to generalize across pixel size
imutil.c: Add medpixi1() and meanpixi1() to handle 8-bit images
imio.c:  Fix 8-bit variables to be unsigned char
ctgread.c: Allow up to 9 magnitudes (/9) instead of 4
catutil.c, polfit.c: Fix xterm computation in polcomp() from + to *

Version 3.8.5 (April 12, 2012)
getdate: Fix string overflow in ts0 assignment (2012-03-27)
gethead: If length of header/file content string is zero, exit with error
sumpix: Improve argument parsing so order doesn't matter (2012-02-01)
matchstar.c: Fix bug in WCSMatch() which miscomputed means
fitsfile.c: Always check first 8 characters of FITS files for "SIMPLE"
imhfile.c: Fix potential overflow in newpixfile assignment (2012-03-27)
imio.c: Change 8-bit pixel values from char to unsigned char

Version 3.8.4 (September 1, 2011)
cphead, keyhead, sethead: Fix overflow bug by increasing size of history from 72 to 128
imgetwcs.c, wcsinit.c, wcs.c, wcs.h, worldpos.c: Add TPV WCS for TAN with PV terms

Version 3.8.3 (August 17, 2011)
getdate: Print hjd in help message instead of jhd
gethead: First column name is FILE_NAME if FILENAME keyword is used (2011-06-29)
gethead: Fix memory leaks (2011-05-19)
gethead: Add -j option to convert RA and Dec to J2000 (2011-05-20)
gethead: Fix -x option for named or unnamed range of extensions (2011-08-12)
gettab: Fix variable name bug so program works!
scat: Fix tab spacing in tmce(2MASS catalog with magnitude errors) (2011-04-21) 
imcat,scat,imwcs,immatch: Fix problem with UCAC3 (2011-04-27)
skycoor: Deal with comma in command line input coordinates (2011-04-19)
skycoor: Drop 99 for skipped RA in input list file (2011-08-17)
hget.c: Free allocated memory in strnsrch() to eliminate memory leak
imhfile.c: Free *newpixname* not pixname. (2011-05-20)
ned_cli.c: Declard *d double, not long double, in ned_getdouble() (2011-05-20)

imrot.c
/* File imrot.c
 * September 25, 2009
 * By Jessica Mink, Harvard-Smithsonian Center for Astrophysics
 * Send bug reports to jmink@cfa.harvard.edu
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <math.h>
#include "libwcs/fitsfile.h"

static void usage();
static void imRot ();
extern char *RotFITS();
extern int DelWCSFITS();

#define MAXFILES 1000
static int maxnfile = MAXFILES;

static int verbose = 0;	/* verbose/debugging flag */
static int mirror = 0;	/* reflect image across vertical axis */
static int automirror = 0;	/* reflect image if IRAF header says to */
static int rotate = 0;	/* rotation in degrees, degrees counter-clockwise */
char outname[128];		/* Name for output image */
static int bitpix = 0;	/* number of bits per pixel (FITS code) */
static int fitsout = 0;	/* Output FITS file from IRAF input if 1 */
static int nsplit = 0;	/* Output multiple FITS files from n-extension file */
static int overwrite = 0;	/* allow overwriting of input image file */
static int version = 0;		/* If 1, print only program name and version */
static int xshift = 0;
static int yshift = 0;
static int shifted = 0;
static int inverted = 0;	/* If 1, invert intensity (-1 * (z-zmax)) */
static int deletewcs = 0;	/* If 1, delete FITS WCS keywords in image */
static int rotatewcs = 1;	/* If 1, rotate FITS WCS keywords in image */
static int extnum = 0;		/* Use extension number instead of EXTNAME for output */
static char *RevMsg = "IMROT WCSTools 3.9.4, 2 August 2016, Jessica Mink (jmink@cfa.harvard.edu)";

int
main (ac, av)
int ac;
char **av;
{
    char *str;
    char c;
    int readlist = 0;
    char *lastchar;
    char filename[128];
    FILE *flist;
    char **fn, *fname;
    char *listfile = NULL;
    char *fni;
    int lfn, i, ifile, nfile;
    double dx, dy;

    nfile = 0;
    fn = (char **)calloc (maxnfile, sizeof(char *));
    outname[0] = 0;

    /* Check for help or version command first */
    str = *(av+1);
    if (!str || !strcmp (str, "help") || !strcmp (str, "-help"))
	usage();
    if (!strcmp (str, "version") || !strcmp (str, "-version")) {
	version = 1;
	usage();
	}

    /* Loop through the arguments */
    for (av++; --ac > 0; av++) {
	str = *av;

	/* List of files to be read */
	if (*str == '@') {
	    readlist++;
	    listfile = ++str;
	    str = str + strlen (str) - 1;
	    av++;
	    ac--;
	    str = str - 1;
	    }

	/* Set image shifts */
	else if (isnum (str)) {
	    if (shifted) {
		dy = atof (str);
		if (dy < 0)
		    yshift = (int) (dy - 0.5);
		else
		    yshift = (int) (dy + 0.5);
		}
	    else {
		shifted = 1;
		dx = atof (str);
		if (dx < 0)
		    xshift = (int) (dx - 0.5);
		else
		    xshift = (int) (dx + 0.5);
		}
	    }

	/* Parameters */
	else if (str[0] == '+') {
	    if (str[1] == 'i')
		setfitsinherit (1);
	    }
	else if (str[0] == '-') {
	    while ((c = *++str) != 0) {
		switch (c) {
 	   	case 'f':	/* FITS file output */
		    fitsout++;
		    break;

		case 'a': /* Flip image around N-S axis if IRAF header says */
		    automirror = 1;
		    break;

		case 'e': /* Delete FITS WCS keywords in image header */
		    deletewcs = 1;
		    break;

		case 'l':	/* image flipped around horizontal axis */
		    mirror = 2;
		    break;

		case 'm':	/* image flipped around vertical axis */
		    mirror = 1;
		    break;

		case 'i':	/* Turn off inheritance from Primary header */
		    setfitsinherit (0);
		    break;

		case 'n':	/* Use extension number instead of EXTNAME */
		    extnum++;
		    break;

		case 'o':	/* Specifiy output image filename */
		    if (ac < 2)
			usage ();
		    if (*(av+1)[0] == '-' || *(str+1) != (char)0)
			overwrite++;
		    else {
			strcpy (outname, *(av+1));
			overwrite = 0;
			av++;
			ac--;
			}
		    break;

		case 'p':	/* Make positive image from negative */
		    inverted++;
		    break;

		case 'r':	/* Rotation angle in degrees */
		    if (ac < 2)
			usage ();
		    rotate = (int) atof (*++av);
		    if (rotate < 0)
			rotate = rotate + 360.0;
		    ac--;
		    break;

		case 's':	/* split input FITS multiextension image file */
		    if (ac < 2)
			usage ();
		    nsplit = atoi (*++av);
		    ac--;
		    break;

		case 'v':	/* more verbosity */
		    verbose++;
		    break;

		case 'w': /* Do not rotate FITS WCS keywords in image header */
		    rotatewcs = 0;
		    break;

		case 'x':	/* Number of bits per pixel */
		    if (ac < 2)
			usage ();
		    bitpix = (int) atof (*++av);
		    ac--;
		    break;

		default:
		    usage();
		    break;
		}
		}
	    }

        /* Image file */
        else if (isfits (str) || isiraf (str)) {
            if (nfile >= maxnfile) {
                maxnfile = maxnfile * 2;
                fn = (char **) realloc ((void *)fn, maxnfile);
                }
            fn[nfile] = str;
            nfile++;
            }

        else {
	    fprintf (stderr,"IMROT: %s is not a FITS or IRAF file \n",str);
            usage();
            }

    }

    /* Process files in file of filenames */
    if (readlist) {
	if ((flist = fopen (listfile, "r")) == NULL) {
	    fprintf (stderr,"IMROT: List file %s cannot be read\n",
		     listfile);
	    usage ();
	    }
	while (fgets (filename, 128, flist) != NULL) {
	    lastchar = filename + strlen (filename) - 1;
	    if (*lastchar < 32) *lastchar = 0;
	    imRot (filename);
	    if (verbose)
		printf ("\n");
	    }
	fclose (flist);
	}

    /* Process files on command line */
    else if (nfile > 0) {
	for (ifile = 0; ifile < nfile; ifile++) {
	    fname = fn[ifile];
	    lfn = strlen (fname);
	    if (lfn < 8)
		lfn = 8;
	    if (nsplit > 0) {
		fni = (char *)calloc (lfn+4, 1);
		for (i = 1; i <= nsplit; i++) {
		    sprintf (fni, "%s,%d", fname, i);
		    if (verbose)
			printf ("%s:\n", fni);
		    imRot (fni);
		    if (verbose)
  			printf ("\n");
		    }
		free (fni);
		}
	    else {
		if (verbose)
		    printf ("%s:\n", fname);
		imRot (fname);
		if (verbose)
  		    printf ("\n");
		}
	    }
	}

    /* If no files processed, print usage */
    else
	usage ();

    return (0);
}

static void
usage ()
{
    fprintf (stderr,"%s\n",RevMsg);
    if (version)
	exit (-1);
    fprintf (stderr,"Shift, Rotate, and/or Reflect FITS and IRAF image files\n");
    fprintf(stderr,"Usage: [-vm][-r rot][-s num] [xshift yshift] file.fits ...\n");
    fprintf(stderr,"  xshift: integer horizontal pixel shift, applied first\n");
    fprintf(stderr,"  yshift: integer vertical pixel shift, applied first\n");
    fprintf(stderr,"  -a: Mirror if IRAF image WCS says to\n");
    fprintf(stderr,"  -e: Delete FITS WCS keywords in image\n");
    fprintf(stderr,"  -f: Write FITS image from IRAF input\n");
    fprintf(stderr,"  -i: Do not append primary header to extension header\n");
    fprintf(stderr,"  -l: Reflect image across horizontal axis\n");
    fprintf(stderr,"  -m: Reflect image across vertical axis\n");
    fprintf(stderr,"  -n: Use extension number instead of EXTNAME for output file name\n");
    fprintf(stderr,"  -o: Allow overwriting of input image, else write new one\n");
    fprintf(stderr,"  -p: Make positive from negative image\n");
    fprintf(stderr,"  -r: Image rotation angle in degrees (default 0)\n");
    fprintf(stderr,"  -s: Split n-extension FITS file\n");
    fprintf(stderr,"  -v: Verbose\n");
    fprintf(stderr,"  -w: Rotate FITS WCS keywords in image\n");
    fprintf(stderr,"  -x: Output pixel size in bits (FITS code, default=input)\n");
    exit (1);
}

static void
imRot (name)
char *name;
{
    char *image;		/* FITS image */
    char *header;		/* FITS header */
    int lhead;			/* Maximum number of bytes in FITS header */
    int nbhead;			/* Actual number of bytes in FITS header */
    int iraffile;		/* 1 if IRAF image */
    char *irafheader = NULL;	/* IRAF image header */
    char newname[256];		/* Name for revised image */
    char *ext = NULL;
    char *newimage;
    char *imext = NULL;
    char *imext1 = NULL;
    char *fname;
    char extname[16];
    int lext = 0;
    int lroot;
    int bitpix0;
    int nx, ny, npix;
    char echar;
    char temp[8];
    char history[64];
    char pixname[256];
    double ctemp;
    double dmax, dpix;
    double bs, bz;

    /* If not overwriting input file, make up a name for the output file */
    if (!overwrite) {
	fname = strrchr (name, '/');
	if (fname)
	    fname = fname + 1;
	else
	    fname = name;
	ext = strrchr (fname, '.');
	if (ext != NULL) {
	    if (!strncmp (ext, ".fit", 4)) {
		if (!strncmp (ext-3, ".ms", 3))
		    ext = ext - 3;
		}
	    lext = (fname + strlen (fname)) - ext;
	    lroot = ext - fname;
	    strncpy (newname, fname, lroot);
	    *(newname + lroot) = 0;
	    }
	else {
	    lext = 0;
	    lroot = strlen (fname);
	    strcpy (newname, fname);
	    }
	imext = strchr (fname, ',');
	imext1 = NULL;
	if (imext == NULL) {
	    imext = strchr (fname, '[');
	    if (imext != NULL) {
		imext1 = strchr (fname, ']');
		*imext1 = (char) 0;
		}
	    }
	}
    else
	strcpy (newname, name);

    /* Open IRAF image */
    if (isiraf (name)) {
	iraffile = 1;
	if ((irafheader = irafrhead (name, &lhead)) != NULL) {
	    if ((header = iraf2fits (name, irafheader, lhead, &nbhead)) == NULL) {
		fprintf (stderr, "Cannot translate IRAF header %s/n",name);
		free (irafheader);
		return;
		}
	    if ((image = irafrimage (header)) == NULL) {
		hgetm (header,"PIXFIL", 255, pixname);
		fprintf (stderr, "Cannot read IRAF pixel file %s\n", pixname);
		free (irafheader);
		free (header);
		return;
		}
	    }
	else {
	    fprintf (stderr, "Cannot read IRAF header file %s\n", name);
	    return;
	    }
	}

    /* Open FITS file */
    else {
	iraffile = 0;
	if ((header = fitsrhead (name, &lhead, &nbhead)) != NULL) {
	    if ((image = fitsrimage (name, nbhead, header)) == NULL) {
		fprintf (stderr, "Cannot read FITS image %s\n", name);
		free (header);
		return;
		}
	    }
	else {
	    fprintf (stderr, "Cannot read FITS file %s\n", name);
	    return;
	    }
	}

    /* Set mirror flag if IRAF WCS is mirrored and automirror is set */
    if (hgetr8 (header, "LTM1_1", &ctemp)) {
	if (ctemp < 0 && automirror)
	    mirror = 1;
	}

    /* Use output filename if it is set on the command line */
    if (outname[0] > 0)
	strcpy (newname, outname);

    /* Create new file name */
    else if (!overwrite) {
	if (imext != NULL) {
	    if (!extnum && hgets (header, "EXTNAME",8,extname)) {
		strcat (newname, "_");
		strcat (newname, extname);
		}
	    else {
		strcat (newname, "_");
		strcat (newname, imext+1);
		}
	    }
	if (inverted)
	    strcat (newname, "p");
	if (shifted)
	    strcat (newname, "s");
	if (mirror == 1)
	    strcat (newname, "m");
	else if (mirror == 2)
	    strcat (newname, "f");
	if (rotate != 0) {
	    strcat (newname, "r");
	    if (rotate < 10 && rotate > -1)
		sprintf (temp,"%1d",rotate);
	    else if (rotate < 100 && rotate > -10)
		sprintf (temp,"%2d",rotate);
	    else if (rotate < 1000 && rotate > -100)
		sprintf (temp,"%3d",rotate);
	    else
		sprintf (temp,"%4d",rotate);
	    strcat (newname, temp);
	    }
	if (bitpix == -64)
	    strcat (newname, "bn64");
	else if (bitpix == -32)
	    strcat (newname, "bn32");
	else if (bitpix == -16)
	    strcat (newname, "bn16");
	else if (bitpix == 32)
	    strcat (newname, "b32");
	else if (bitpix == 16)
	    strcat (newname, "b16");
	else if (bitpix == 8)
	    strcat (newname, "b8");
	if (fitsout)
	    strcat (newname, ".fits");
	else if (lext > 0) {
	    if (imext != NULL) {
		echar = *imext;
		*imext = (char) 0;
		strcat (newname, ext);
		*imext = echar;
		if (imext1 != NULL)
		    *imext1 = ']';
		}
	    else
		strcat (newname, ext);
	    }
	}
    else
	strcpy (newname, name);

    if (verbose) {
	fprintf (stderr,"%s\n",RevMsg);
	fprintf (stderr,"Shift, rotate, and/or reflect ");
	if (iraffile)
	    fprintf (stderr,"IRAF image file %s", name);
	else
	    fprintf (stderr,"FITS image file %s", name);
	fprintf (stderr, " -> %s\n", newname);
	}

    if (bitpix != 0) {
	hgeti4 (header, "BITPIX", &bitpix0);
	if (verbose)
	    fprintf (stderr, "IMROT: %d bits/pixel -> %d bits/pixel\n",
		     bitpix0, bitpix);
	sprintf (history, "New copy of %s BITPIX %d -> %d",
		 name, bitpix0, bitpix);
	hputc (header,"HISTORY",history);
	}

    /* Make positive image from negative if requested */
    if (inverted) {
	bitpix = 16;
	hgeti4 (header, "BITPIX", &bitpix0);
	bs = 1.0;
	hgetr8 (header, "BSCALE", &bs);
	bz = 0.0;
	hgetr8 (header, "BZERO", &bz);
	nx = 1;
	hgeti4 (header, "NAXIS1", &nx);
	ny = 1;
	hgeti4 (header, "NAXIS2", &ny);
	npix = nx * ny;
	dmax = -maxvec (image, bitpix0, bz, bs, 0, npix);
	addvec (image, bitpix0, bz, bs, 0, npix, dmax);
	dpix = -1.0;
	multvec (image, bitpix0, bz, bs, 0, npix, dpix);
	}

    /* Delete FITS WCS keywords in the image if requested */
    if (deletewcs) {
	DelWCSFITS (header, verbose);
	}

    if ((newimage = RotFITS (name,header,image,xshift,yshift,rotate,mirror,
			     bitpix,rotatewcs,verbose)) == NULL) {
	fprintf (stderr,"Cannot rotate image %s; file is unchanged.\n",name);
	}
    else {
	if (bitpix != 0)
	    hputi4 (header, "BITPIX", bitpix);
	if (iraffile && !fitsout) {
	    if (irafwimage (newname,lhead,irafheader,header,newimage) > 0) {
		if (verbose)
		    printf ("%s: written successfully.\n", newname);
		else
		    printf ("%s\n", newname);
		}
	    else if (verbose)
		printf ("IMROT: File %s not written.\n", newname);
	    }
	else {
	    if (fitswimage (newname, header, newimage) > 0) {
		if (verbose)
		    printf ("%s: written successfully.\n", newname);
		else
		    printf ("%s\n", newname);
		}
	    else if (verbose)
		printf ("IMROT: File %s not written.\n", newname);
	    }
	free (newimage);
	}

    free (header);
    if (iraffile)
	free (irafheader);
    free (image);
    return;
}
/* Apr 15 1996	New program
 * Apr 18 1996	Add option to write to current working directory
 * May  2 1996	Pass filename to rotFITS
 * May 28 1996	Change rotFITS to RotFITS
 * Jun  6 1996	Always write to current working directory
 * Jun 14 1996	Use single image buffer
 * Jul  3 1996	Allow optional overwriting of input image
 * Jul 16 1996	Update header reading and allocation
 * Aug 26 1996	Change HGETC call to HGETS; pass LHEAD in IRAFWIMAGE
 * Aug 27 1996	Remove unused variables after lint
 *
 * Feb 21 1997  Check pointers against NULL explicitly for Linux
 * Aug 13 1997	Fix bug when overwriting an image
 * Dec 15 1997	Add capability of reading and writing IRAF 2.11 images
 *
 * Feb 24 1998	Add ext. to filename if writing part of multi-ext. file
 * May 26 1998	Fix bug when writing .imh images
 * May 27 1998	Include fitsio.h instead of fitshead.h
 * Jul 24 1998	Make irafheader char instead of int
 * Aug  6 1998	Change fitsio.h to fitsfile.h
 * Oct 14 1998	Use isiraf() to determine file type
 * Nov 30 1998	Add version and help commands for consistency
 * Jun  8 1999  Return image pointer from RotFITS, not flag
 * Oct 22 1999	Drop unused variables after lint
 *
 * Jan 24 2000	Add to name if BITPIX is changed
 * Mar 23 2000	Use hgetm() to get the IRAF pixel file name, not hgets()
 * Jul 20 2000	Use .fits, not .fit, extension on output FITS file names
 * Jul 20 2000	Add -s option to split multi-extension FITS files
 * Sep 12 2000	Echo new file name to standard output, if not verbose
 * Sep 12 2000	Use .extname if multi-extension extraction, not _number
 *
 * Jan 18 2001	Add automirror -a option
 * Feb 13 2001	Add -o name option to -o argument (from imwcs)
 *
 * Jan 28 2004	Add option to shift file data within file (before rotating)
 * May  6 2004	Add -i to avoid appending primary header to extension header
 * Jun 14 2004	Write HISTORY only if BITPIX is changed
 *
 * Aug 17 2005	If rotation angle < 360, add 360
 * Aug 17 2005	-m replaces -l for mirror reflection, -l flips
 * Aug 18 2005	Add -p to make positive image from negative
 *
 * Jan 19 2006	Add +i to force inheritance of keywords from primary header
 * Jun 21 2006	Clean up code
 * Jun 22 2006	Check for two-token extension .ms.fit(s)
 *
 * Jan  5 2007	Fix BSCALE and BZERO hget calls
 *
 * Apr  4 2007	Add -w option to not rotate and -e option to erase WCS keywords
 *
 * Mar 27 2009	Use _ instead of . to separate extension name or number in output filename
 * Mar 27 2009	Add -n option to force use of extension number instead of EXTNAME
 * Sep 25 2009	Declare DelWCSFITS()
 */
